src/rust/lib_ccxr/src/encoder/txt_helpers.rs

```diff
@@ -0,0 +1,483 @@
+use crate::info;
+use crate::util::encoding::*;
+use std::cmp;
+use std::convert::TryFrom;
+
+pub enum EncoderError {
+    Retry = -100,           // CCX_EAGAIN
+    EOF = -101,             // CCX_EOF
+    InvalidArgument = -102, // CCX_EINVAL
+    Unsupported = -103,     // CCX_ENOSUPP
+    OutOfMemory = -104,     // CCX_ENOMEM
+}
+fn find_limit_characters(
+    line: &[u8],
+    first_non_blank: &mut i32,
+    last_non_blank: &mut i32,
+    max_len: usize,
+) {
+    // initialize to -1 (meaning “not found”)
+    *first_non_blank = -1;
+    *last_non_blank = -1;
+
+    // clamp to avoid out‑of‑bounds
+    let limit = cmp::min(line.len(), max_len);
+
+    for (i, &c) in line.iter().take(limit).enumerate() {
+        // detect non‑blank
+        if c != b' ' && c != 0x89 {
+            if *first_non_blank < 0 {
+                *first_non_blank = i as i32;
+            }
+            *last_non_blank = i as i32;
+        }
+        // break on end‑of‑string or newline
+        if c == b'\0' || c == b'\n' || c == b'\r' {
+            break;
+        }
+    }
+}
```

Comment:
You may use rust logic to get last_non_blank and first_non_blank by positions iter and other things, making it native rust.